### PR TITLE
feat(compass-sidebar): unified sidebar experience for multiple connections - COMPASS-8016

### PR DESCRIPTION
**Pull request by himanshusinghs**
## Description
With this PR we're unifying the active connections panel and saved connections panel into one single panel "Connections" panel.

### Todo

- [x] Merge menu items for both connected and disconnected connections and have them render (maybe disabled) based on the connection status
- [x] Disable Edit connection menu item when connection is connected
- [x] Ensure that the menu is wide and big enough to accommodate the action items
- [x] Ensure that navigation with keyboard works
- [x] Fix the flickering connection item problem when the connection status changes
- [x] Add collapse all button in the connection lists header
- [x] Add new connection button in the connections list header
- [x] Adapt existing tests for the new sidebar
- [x] Add empty state for connections list (CTA to add new connection)
- [x] Add inactive state of connections to connections navigation tree
- [x] Add new tests for the new sidebar
- [x] Disable Connect menu item when active connections are more than maximum number of active connections
- [x] Ensure that tabbing in connection tree is able to identify the action buttons and brings them into focus when tabbed again
- [x] Check why switching between tabs does not change the active item in the sidebar
- [x] Check why collapse all does not work for freshly connected connection

### Demo

https://github.com/mongodb-js/compass/assets/10037761/25845ba0-fd91-4f15-b8bd-bdf4a0e4591b

### Notes for reviewer:
- Start with changes in compass-components because there we've only added an icon component
- Move onto compass package because the changes here ensures that the sidebar does not remount all the time there is an irrelevant change in workspace.
- Then review compass-connections because here we've added a hook that forms the base for us to render all the connections in the new sidebar
- In compass-connections-navigation the important changes are:
  - The padding that we're adjusting to align the sidebar items [here](https://github.com/mongodb-js/compass/pull/5877/files#diff-58ff0015bfeec80e48671a55a1efc6750769a8fcecf3513d62d02b3232dbae05R9-R11).
  - Changes in navigation item to support having the status markers and a few styling changes to make them round around borders
  - Changes in item-actions to support unified menu items for connections
- then in compass-sidebar
  - ignore the removed components, I will make sure to cover the test cases that we've removed
  - focus primarily on connections-navigation component, you might wanna recommend some better way for me to do the [following](https://github.com/mongodb-js/compass/pull/5877/files#diff-db431982e35eca0904f03ec50fc20d62f9d7d52b95f769f5278264fd38a2b6adR144-R186).

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-sidebar/src/components/multiple-connections/connections-navigation.tsx [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+}) => {
+  const {
+    openShellWorkspace,
+    openPerformanceWorkspace,
+    openDatabasesWorkspace,
+    openCollectionsWorkspace,
+    openCollectionWorkspace,
+    openEditViewWorkspace,
+  } = useOpenWorkspace();
+  const connections = useMemo(() => {
```
Comment (himanshusinghs) [inline packages/compass-sidebar/src/components/multiple-connections/connections-navigation.tsx:144]: So about this memo - it could have been in the mapState but I opted for memo in the hopes of avoiding re-calculation on every irrelevant change in the sidebar store.
I understand that it means I need to pass down the entire database, instance and isPerformanceTabSupported slice to this component but the other option was to refactor the store a little and perform this calculation within the store.

But to be honest I also did not spend too much time on this so please feel free to recommend something for this. I am open to suggestions.

File: packages/compass/src/app/components/workspace.tsx [hunk 2]
```diff
@@ -67,6 +67,29 @@ export default function Workspace({
   const multiConnectionsEnabled = usePreference(
     'enableNewMultipleConnectionSystem'
   );
+  const renderSidebar = useCallback(
+    () => (
+      <CompassSidebarPlugin
+        singleConnectionConnectionInfo={singleConnectionConnectionInfo}
+      />
+      // This will be gone once we merge my-queries PR
+      // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (gribnoysup) [inline packages/compass/src/app/components/workspace.tsx:76]: This is a scary way to break the rules of hooks especially when connection info is what's being passed here. Why is this required?

Comment (himanshusinghs) [reply packages/compass/src/app/components/workspace.tsx:76]: this was a left over and I was supposed to remove this, thank you for catching up on this.

File: packages/compass-connections-navigation/src/item-actions.ts [hunk 3]
```diff
@@ -50,23 +57,25 @@ export const notConnectedConnectionItemActions = ({
...
-  isFavorite: boolean;
   isPerformanceTabSupported: boolean;
 }): NavigationItemActions => {
-  const actions: NavigationItemActions = [];
-  if (!isReadOnly) {
-    actions.push({
+  const connectionManagementActions = notConnectedConnectionItemActions({
+    connectionInfo,
+    isEditDisabled: true,
+    editDisabledDescription: 'Cannot edit an active connection',
```
Comment (lerouxb) [inline packages/compass-connections-navigation/src/item-actions.ts:71]: I'm wondering if it really is so bad if we just remove the option from the list rather.

File: packages/compass-connections-navigation/src/item-actions.ts [hunk 4]
```diff
@@ -84,23 +93,19 @@ export const connectedConnectionItemActions = ({
...
       variant: 'destructive',
-    }
-  );
+    },
+    { separator: true },
+    ...connectionManagementActions,
+  ];
+
+  if (isReadOnly) {
+    actions.splice(0, 1);
```
Comment (lerouxb) [inline packages/compass-connections-navigation/src/item-actions.ts:107]: What's being removed here?

Comment (himanshusinghs) [reply packages/compass-connections-navigation/src/item-actions.ts:109]: sorry for the overlooking here - i added [a comment](https://github.com/mongodb-js/compass/pull/5877/commits/cd65815340e0839354c9cff6486fcc10af3e6b9c) describing what is getting removed and why.

**Review by lerouxb — APPROVED** [review]

File: packages/compass/src/app/components/workspace.tsx (hunk 2 above)
Comment (himanshusinghs) [reply packages/compass/src/app/components/workspace.tsx]: We removed these changes now

File: packages/compass-connections/src/hooks/use-active-connections.ts
Comment (himanshusinghs) [inline packages/compass-connections/src/hooks/use-active-connections.ts:1]: I basically moved most of the logic in this file to use-connections-with-status and then using that hook to get the active connections.

Comment (himanshusinghs) [inline packages/compass-connections/src/hooks/use-active-connections.ts:1]: I basically moved most of the logic in this file to use-connections-with-status and then using that hook to get the active connections.
